packages/openapi-v3/src/__tests__/unit/json-to-schema.unit.ts

```diff
@@ -139,6 +140,135 @@ describe('jsonToSchemaObject', () => {
     expect(jsonToSchemaObject(inputDef)).to.eql(expectedDef);
   });
 
+  it('handles circular references with $ref', () => {
+    const schemaJson: JsonSchema = {
+      title: 'ReportState',
+      properties: {
+        // ReportState[]
+        states: {type: 'array', items: {$ref: '#/definitions/ReportState'}},
+        benchmarkId: {type: 'string'},
+        color: {type: 'string'},
+      },
+    };
+    const schema = jsonToSchemaObject(schemaJson);
+    expect(schema).to.eql({
+      title: 'ReportState',
+      properties: {
+        states: {
+          type: 'array',
+          items: {$ref: '#/components/schemas/ReportState'},
+        },
+        benchmarkId: {type: 'string'},
+        color: {type: 'string'},
+      },
+    });
+  });
+
+  it('handles circular references with object', () => {
+    const schemaJson: JsonSchema = {
+      title: 'ReportState',
+      properties: {
+        benchmarkId: {type: 'string'},
+        color: {type: 'string'},
+      },
+    };
+    // Add states: ReportState[]
+    schemaJson.properties!.states = {type: 'array', items: schemaJson};
+    const schema = jsonToSchemaObject(schemaJson);
+    expect(schema).to.eql({
+      title: 'ReportState',
+      properties: {
+        states: {
+          type: 'array',
+          items: {$ref: '#/components/schemas/ReportState'},
+        },
+        benchmarkId: {type: 'string'},
+        color: {type: 'string'},
+      },
+    });
+  });
+
+  it('handles indirect circular references with $ref', () => {
+    const schemaJson: JsonSchema = {
+      title: 'ReportState',
+      properties: {
+        parentState: {$ref: '#/definitions/ParentState'},
+        benchmarkId: {type: 'string'},
+        color: {type: 'string'},
+      },
+      definitions: {
+        ParentState: {
+          title: 'ParentState',
+          properties: {
+            timestamp: {type: 'string'},
+            state: {$ref: '#/definitions/ReportState'},
+          },
+        },
+      },
+    };
+    const schema = jsonToSchemaObject(schemaJson);
+    expect(schema).to.eql({
+      title: 'ReportState',
+      properties: {
+        parentState: {$ref: '#/components/schemas/ParentState'},
+        benchmarkId: {type: 'string'},
+        color: {type: 'string'},
+      },
+      definitions: {
+        ParentState: {
+          title: 'ParentState',
+          properties: {
+            timestamp: {type: 'string'},
+            state: {$ref: '#/components/schemas/ReportState'},
+          },
+        },
+      },
+    });
+  });
+
+  it('handles indirect circular references with object', () => {
+    const parentStateSchema: JSONSchema6Definition = {
```

Comment:
Why do we name it as `JSONSchema_6_Definition`?

Reply:
It's based on JSON Schema draft 6 - https://json-schema.org/draft-06/json-schema-release-notes.html

Reply:
It makes me wonder - why do we need `JSONSchema6Definition` interface here and cannot use `JsonSchema` that we are using in all other places?

AFAICT:

- `JSONSchema6Definition` is defined as `JSONSchema6 | boolean`
- `JsonSchema` is our alias for `JSONSchema6`

So IIUC, the difference between `JSONSchema6Definition` and `JsonSchema` is that the former also allows naked boolean values (`parentSchema = true`) while the latter requires an object.

Do you really need the `| boolean` part here?